### PR TITLE
150 make selfplay pong work

### LINKED ISSUE #150: Make selfplay pong work
**Goal**
Train an AI agent using PPO with replay buffer to achieve good performance on the Pong game in the Petting Zoo environment

**Acceptance Criteria**
- [ ] Trained AI agent is able to play Pong game effectively, achieving an averaged score higher than [3000 steps](https://wandb.ai/openrlbenchmark/openrlbenchmark/reports/Pettingzoo-s-Multi-agent-Atari-CleanRL-s-PPO--VmlldzoyMDkxNTE5)
- [ ] PPO algorithm is successfully integrated into the cogment-verse training pipeline
- [ ] Write a detailed README file that includes instructions on how to run the code and how to reproduce the results

### PR DESCRIPTION
**Pull request by lhnguyen102**
**Description**
This PR implements Asynchronous Proximal Policy Optimization (APPO) for the Petting Zoo pong environment. The objective is to achieve performance levels that are comparable to those mentioned in the literature.

**Related Issue**
#150 

**Change Made**
- [x] Implement APPO sample producer for self-training, human demo, and human feedback
- [x] Add model registry to `SampleProducerSession` in order to compute the values and log-likelihood for a rollout 
- [x] Add rollout buffer as well as  replay buffer for APPO
- [x] Add jupyter notebook associated with the python scripts that enable running the Cog-Verse framework using SageMaker i.e. `cloud`
- [x] Refactor `update_parameters` and `compute_gae`
- [x] Refactor `HumanReplayBuffer` in order to handle the addition of multiple observations at once
- [x] Enforce the code by adding type hint in order to increase the code tractability in VS code

**Additional Notes**
- Performance of APPO on Pong (average reward ~2000) is slightly improved compared to the previous PPO version but it could not reach the same performance that we aim (~4000)
- I decided to merge the current branch to main eventhough the performance issue are not resolved because we'll need all these features to be ready for AAMAS 2023 on May 29

closes #150

### CODE REVIEW

File: config/experiment/ppo_atari_pz/observe_pong_pz.yaml [hunk 1]
```diff
@@ -4,15 +4,28 @@ defaults:
...
+    #   agent_config:
+    #     model_id: compassionate_faraday_0_model
+    #     model_iteration: -1
+
     - name: first_0
       implementation: actors.ppo_atari_pz.PPOActor
       agent_config:
-        model_id: thirsty_matsumoto_0_model
-        model_iteration: 351
+        model_id: nostalgic_pare_0_model
```
Comment (wduguay-air) [inline config/experiment/ppo_atari_pz/observe_pong_pz.yaml:24]: Switch those model ids for `???`

**Review by wduguay-air — COMMENTED** [review]
First pass of review

File: config/experiment/ppo_atari_pz/observe_pong_pz.yaml (hunk 1 above)
Comment (lhnguyen102) [reply config/experiment/ppo_atari_pz/observe_pong_pz.yaml:24]: Done

File: actors/ppo_atari_pz.py [hunk 2]
```diff
@@ -291,84 +337,106 @@ async def trial_sample_sequences_producer_impl(self, sample_producer_session: Sa
...
+
+            # Save episode reward i.e., number of total steps for an episode
+            step += 1
+            total_reward += 1
+            if trial_done:
+                episode_rewards.append(torch.tensor(total_reward / num_players, dtype=self._dtype))
+                total_reward = 0
+
+            # Produce sample for training task
+            if step % self._cfg.num_rollout_steps == 0 or trial_done:
```
[the remark this answers is not in the document]
Comment (lhnguyen102) [reply actors/ppo_atari_pz.py:408]: We only want to produce after N steps of rollout, In this case will be 128 (i.e., `num_rollout_steps: 128`). This is one of a feature of PPO algorithm (see [Algorithm 1](https://arxiv.org/pdf/1707.06347.pdf)). The reason is that the number of steps of a pong trial is >3000 steps. We want to update the algorithm each small N steps in order to avoid the divergence in performance. For example, in the case of Pong, 128 is commonly used. For hopper, it is 2048

File: actors/ppo_atari_pz.py [hunk 3]
```diff
@@ -291,84 +337,106 @@ async def trial_sample_sequences_producer_impl(self, sample_producer_session: Sa
...
+            if not trial_done:
+                action_value = torch.tensor(
+                    player_action_space.deserialize(actor_sample.action).value, dtype=self._dtype
+                )
+                current_players.append(player_actor_name)
 
-        # Keeping the samples grouped by trial by emitting only one grouped sample at the end of the trial
-        sample_producer_session.produce_sample((observation, action, reward, done, current_players))
+            # Compute values and log probs
+            if step % self._cfg.num_rollout_steps < self._cfg.num_rollout_steps and not trial_done:
```
[the remark this answers is not in the document]
Comment (lhnguyen102) [reply actors/ppo_atari_pz.py:389]: you're right!. Here is the correction `step % self._cfg.num_rollout_steps > 0 and not trial_done`. I want to collect only 128 samples and ignore the 129th, but I will need the 129th to compute GAE

File: actors/ppo_atari_pz.py [hunk 4]
```diff
@@ -512,22 +570,76 @@ def normalize_rewards(self, rewards: list, model: PPOModel) -> list:
...
+                gae = delta + gamma * lambda_ * (1 - dones[i]) * gae
                 advs.append(gae)
+                next_value = values[i]
         advs.reverse()
-        return torch.vstack(advs)
+
+        return advs
+
+
+class RolloutBuffer:
```
[the remark this answers is not in the document]
Comment (lhnguyen102) [reply actors/ppo_atari_pz.py:598]: It needs to be separate because `replay_buffer` is fundamentally differently than rollout buffer. Replay buffer means that a sample in this buffer can be reused more than 1 while a sample in rollout buffer can only be used once. In addition, rollout buffer does not have `sample` method and `self.next_observation`. So, I decided to separate them to avoid the confusion

File: actors/ppo_atari_pz.py [hunk 5]
```diff
@@ -266,22 +307,27 @@ def __init__(self, environment_specs: EnvironmentSpecs, cfg: EnvironmentConfig)
...
+    @staticmethod
+    def is_time_to_load_model(sample_tick_id: int, num_rollout_steps: int, trial_done: bool) -> bool:
+        """Load model from model registry if it matches predefined conditions regarding
+        trial status and tick id"""
+        is_multiple_of_rollout = (sample_tick_id + 1) % num_rollout_steps == 0
+        is_not_trial_done = not trial_done
+        is_first_tick = sample_tick_id == 0
+        is_loaded = is_multiple_of_rollout and is_not_trial_done
 
+        return is_loaded or is_first_tick
```
[the remark this answers is not in the document]
Comment (lhnguyen102) [reply actors/ppo_atari_pz.py:327]: Thanks!. It is been replaced as you suggested

File: cogment_verse/replay_buffers/torch_replay_buffer.py [hunk 6]
```diff
@@ -78,11 +178,11 @@ def sample(self, num):
             indices = self._rng.choice(self.size(), size=num, replace=False)
 
         return TorchReplayBufferSample(
-            observation=self.observation[indices],
-            next_observation=self.next_observation[indices],
-            action=self.action[indices],
-            reward=self.reward[indices],
-            done=self.done[indices],
```
[the remark this answers is not in the document]
Comment (lhnguyen102) [reply cogment_verse/replay_buffers/torch_replay_buffer.py:85]: Thank you for reminding me this. I need to do the same for other buffer. It is because of the safety precaution. I think they store at the same memory pointer. It would be bad if we do any manipulation with this observation. You can test it using this function. 
```python
import torch
def samestorage(x,y):
    if x.storage().data_ptr()==y.storage().data_ptr():
        print("same storage")
    else:
        print("different storage")
```

File: cogment_verse/replay_buffers/human_data_buffer.py [hunk 7]
```diff
@@ -82,12 +82,25 @@ def add(self, observation: np.ndarray, action: np.ndarray, feedback: Union[np.nd
...
     def sample(self, num_samples: int) -> Tuple[np.ndarray, np.ndarray]:
         """Select randomly the samples"""
+        np.random.seed(self.seed + self.count)
         if self.num_total < num_samples:
             indices = range(self.num_total)
         else:
-            indices = self._rng.choice(self.num_total, size=num_samples, replace=False)
+            indices = np.random.choice(self.num_total, size=num_samples, replace=False)
 
         return (self.observations[indices], self.actions[indices])
```
[the remark this answers is not in the document]
Comment (lhnguyen102) [reply cogment_verse/replay_buffers/human_data_buffer.py:105]: It is just for offline analysis. I reserved it for the TAMER algorithm.  You're right. I should learn how to use `trial_datastore` (this why I need your notebook). I will keep it for now. When I am able to implement TAMER, i will refactor the code accordingly

File: cogment_verse/replay_buffers/human_data_buffer.py [hunk 8]
```diff
@@ -67,7 +67,7 @@ def loading(self, value: bool) -> None:
                 self.feedback = np.zeros((self.capacity, 1), dtype=np.float32)
             self._ptr = 0
             self.num_total = 0
-            self._rng = np.random.default_rng(self.seed)
+            self.count = 0
```
[the remark this answers is not in the document]
Comment (lhnguyen102) [reply cogment_verse/replay_buffers/human_data_buffer.py:70]: It is nothing wrong with it. I read [this post](https://pytorch.org/docs/stable/notes/randomness.html) about reproducibility which I want in the case to debug APPO. I've tested on APPO algorithm, it is actually correct. I also change the seed in line 103 for each sample selection

**Review by wduguay-air — APPROVED** [review]
